services/src/main/scala/cromwell/pipeline/service/AuthService.scala

```diff
@@ -26,16 +20,22 @@ class AuthService(userRepository: UserRepository, authUtils: AuthUtils)(implicit
 
   import authUtils._
 
-  def signIn(request: SignInRequest): Future[Option[AuthResponse]] =
-    OptionT(userRepository.getUserByEmail(request.email))
-      .filter(user => user.passwordHash == StringUtils.calculatePasswordHash(request.password.value, user.passwordSalt))
-      .map { user =>
-        val accessTokenContent = AccessTokenContent(user.userId)
-        val refreshTokenContent = RefreshTokenContent(user.userId, None)
-        getAuthResponse(accessTokenContent, refreshTokenContent, Instant.now.getEpochSecond)
+  def signIn(request: SignInRequest): Future[Option[AuthResponse]] = responseFromOptUser(authFilter(request))
+
+  def authFilter(request: SignInRequest): OptionT[Future, User] =
+    OptionT(userRepository.getUserByEmail(request.email)).semiflatMap[User] { user =>
+      if (passwordCorrect(request, user)) Future(user)
+      else {
+        Future.failed(IncorrectPasswordException("Wrong password"))
       }
-      .value
-      .map(_.flatten)
+    }
+
+  def responseFromOptUser(optUser: OptionT[Future, User]): Future[Option[AuthResponse]] =
```

Comment:
It may be nitpicking, but this code doesn't look like Scala, despite the fact that you are using cats here.
We love `.map`, `flatMap` etc, since they allow us to abstract over the type of the container (such as `Option`, `OptionT`) and use one function with multiple containers without any problem. You can't reuse your `def responseFromOptUser` with any other type of container (even if it is a simple `Option`).
It would be better to change the `responseFromOptUser` method as follows.
`def responseFromUser(user: User): Option[AuthResponse]`
This method shouldn't care about addtional wrappers like `OptionT`. If you give him a user, he will give a response. That's all.
In that case, `signIn` method should be rewritten as follows
```
def signIn(request: SignInRequest): Future[Option[AuthResponse]] =
  authFilter(request).subflatMap(responseFromUser).value
```
Also, the name of the method `authFilter`  does not correspond to what it does. It takes `SignInRequest` and returns `User`. According to the name, it should only filter something. According to the types, it somehow transforms `request` into `user`.
And don't forget that method names should be verbs.

Reply:
1) "nitpicking", - no, it is not the case!
2) your advice a pretty cool, I suppose I just need more cats exp to observe all possible variants!